dask_cuda/tests/test_spill.py

```diff
@@ -143,9 +157,147 @@ def get_data(worker, total_size):
             assert (abs(res / x.size) - 0.5) < 1e-3
 
             await client.run(get_data, cluster.workers[0].Worker, x.nbytes)
+            host_chunks = await client.run(lambda: len(get_worker().data.host))
+            disk_chunks = await client.run(lambda: len(get_worker().data.disk))
+            for hc, dc in zip(host_chunks.values(), disk_chunks.values()):
+                if params["spills_to_disk"]:
+                    assert dc > 0
+                else:
+                    assert hc > 0
+                    assert dc == 0
+
+
+@pytest.mark.parametrize(
+    "params",
+    [
+        {
+            "device_memory_limit": 1e9,
+            "memory_limit": 4e9,
+            "host_target": 0.6,
+            "host_spill": 0.7,
+            "spills_to_disk": False,
+        },
+        {
+            "device_memory_limit": 1e9,
+            "memory_limit": 2e9,
+            "host_target": 0.6,
+            "host_spill": 0.6,
+            "spills_to_disk": True,
+        },
+    ],
+)
+def test_cudf_device_spill(params):
+    @gen_cluster(
+        client=True,
+        ncores=[("127.0.0.1", 1)],
+        Worker=Worker,
+        timeout=300,
+        worker_kwargs={
+            "memory_limit": params["memory_limit"],
+            "data": DeviceHostFile(
+                device_memory_limit=params["device_memory_limit"],
+                memory_limit=params["memory_limit"],
+            ),
+        },
+        config={
+            "distributed.comm.timeouts.connect": "20s",
+            "distributed.worker.memory.target": params["host_target"],
+            "distributed.worker.memory.spill": params["host_spill"],
+        },
+    )
+    def test_device_spill(client, scheduler, worker):
+
+        cdf = dask.datasets.timeseries(
+            dtypes={"x": int, "y": float}, freq="30ms"
+        ).map_partitions(cudf.from_pandas)
+
+        sizes = yield client.compute(cdf.map_partitions(lambda df: df.__sizeof__()))
+        sizes = sizes.tolist()
+        nbytes = sum(sizes)
+        part_index_nbytes = (yield client.compute(cdf.partitions[0].index)).__sizeof__()
+
+        cdf2 = cdf.persist()
+        yield wait(cdf2)
+
+        del cdf
+
+        yield client.run(worker_assert, nbytes, 32, 2048 + part_index_nbytes)
+
+        host_chunks = yield client.run(lambda: len(get_worker().data.host))
+        disk_chunks = yield client.run(lambda: len(get_worker().data.disk))
```

Comment:
It's not important, but because you're using normal `Worker`s here rather than `Nanny`s, have direct access to the workers here.  You can look at `worker.data.host` and `worker.data.disk` directly, without using `client.run`.

Reply:
Yes, but I prefer to call `client.run` here to make sure that works and to make all tests consistent, I've added some more `client.run` now. Perhaps we can generalize that in the future to a single assertion function.